File: 2024-09-Septmber-GFG POTD/September 27 - K sized subarray maximum/Solution.py
# { Driver Code Starts
from collections import deque

class Solution:
    # Function to find maximum of each subarray of size k.
    def max_of_subarrays(self, arr, n, k):
        result = []
        dq = deque()

        for i in range(n):
            # Remove elements not part of this window (out of bounds for subarray)
            if dq and dq[0] <= i - k:
                dq.popleft()

            # Remove elements from the deque that are smaller than the current element,
            # because they will not be the maximum for any future subarray.
            while dq and arr[dq[-1]] <= arr[i]:
                dq.pop()

            # Add the current element index at the back of the deque
            dq.append(i)

            # The maximum of the current window is at the front of the deque
            if i >= k - 1:
                result.append(arr[dq[0]])

        return result


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n, k = map(int, input().split())
        arr = list(map(int, input().split()))
        ob = Solution()
        res = ob.max_of_subarrays(arr, n, k)
        print(" ".join(map(str, res)))
# } Driver Code Ends

File: 2024-09-Septmber-GFG POTD/September 27 - K sized subarray maximum/Solution.java
//{ Driver Code Starts
// Initial template for JAVA

import java.io.*;
import java.lang.*;
import java.util.*;

public class Main {
    public static void main(String args[]) throws IOException {
        // taking input using class Scanner
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            // taking total number of elements
            int k = Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;
            ArrayList<Integer> res = new Solution().max_of_subarrays(k, arr);

            // printing the elements of the ArrayList
            for (int i = 0; i < res.size(); i++) System.out.print(res.get(i) + " ");
            System.out.println();
        }
    }
}
// } Driver Code Ends


// User function template for JAVA

class Solution {
    // Function to find maximum of each subarray of size k.
    public ArrayList<Integer> max_of_subarrays(int k, int arr[]) {
        // Your code here
        ArrayList<Integer> result = new ArrayList<>();
        Deque<Integer> deque = new LinkedList<>();

        for (int i = 0; i < arr.length; i++) {
            // Remove elements not part of this window (out of bounds for subarray)
            if (!deque.isEmpty() && deque.peekFirst() <= i - k) {
                deque.pollFirst();
            }

            // Remove elements from the deque that are smaller than the current element,
            // because they will not be the maximum for any future subarray.
            while (!deque.isEmpty() && arr[deque.peekLast()] <= arr[i]) {
                deque.pollLast();
            }

            // Add the current element index at the back of the deque
            deque.offerLast(i);

            // The maximum of the current window is at the front of the deque
            if (i >= k - 1) {
                result.add(arr[deque.peekFirst()]);
            }
        }

        return result;
    }
}